pandaproxy: add max memory check for incoming requests

IoannisRP opened the pull request:
fixes: [CORE-8335](https://redpandadata.atlassian.net/browse/CORE-8335)

In the pandaproxy server, if a request comes in that is larger than the total available memory, every other request is blocked. A test is added to make sure that a request bigger than the available memory returns an error.

For better monitoring, the following metrics have been added,
prefixed with `[vectorized|redpanda]_[rest_proxy|schema_registry]_` :
| Metric | Type | Description | Labels | Aggregation labels |
| -- | -- | -- | -- | -- |
| `inflight_requests_usage_ratio` | gauge | Usage ratio of in-flight requests in the [rest_proxy\|schema_registry] | `shard` |  |
| `inflight_requests_memory_usage_ratio` | gauge | Memory usage ratio of in-flight requests in the [rest_proxy\|schema_registry]  in bytes | `shard` |  |
| `queued_requests_memory_blocked` | gauge | Number of requests queued in [rest_proxy\|schema_registry], due to memory limitations | `shard` |  |

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.3.x
- [x] v24.2.x
- [x] v24.1.x

## Release Notes

### Improvements

- Added metrics for pandaproxy resource usage. 

[CORE-8335]: https://redpandadata.atlassian.net/browse/CORE-8335?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
pandaproxy: add max memory check for incoming requests

IoannisRP commented:
/dt

vbotbuildovich commented:
<!--NEW FAILURES-->
<!--1c438361-58b2-42a8-a808-7d1dd8a1f578-->
<!--body-->
non flaky failures in https://buildkite.com/redpanda/redpanda/builds/59616#0193b7b4-8c87-4047-b52a-c85d1f6a6b2f:
```
"rptest.tests.datalake.compaction_gaps_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3"
```
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59616-->
<!--fcd3882c-faf2-4358-8233-9b306b65f4da-->
<!--title-->
## Retry command for [Build#59616](https://buildkite.com/redpanda/redpanda/builds/59616)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/compaction_gaps_test.py::CompactionGapsTest.test_translation_no_gaps@{"cloud_storage_type":1}
```
<!--/body-->

BenPope commented on src/v/pandaproxy/reply.h:
```diff
@@ -65,6 +65,11 @@ inline ss::http::reply& set_reply_too_many_requests(ss::http::reply& rep) {
 
+inline ss::http::reply& set_reply_payload_too_large(ss::http::reply& rep) {
+    return rep.set_status(ss::http::reply::status_type::payload_too_large)
+      .add_header("Retry-After", "0");
```
Retry isn't useful here, I wouldn't return any headers,

IoannisRP committed:
pandaproxy: add semamphore usage metrics

BenPope commented on src/v/pandaproxy/server.cc:
```diff
@@ -249,4 +253,69 @@ ss::future<> server::stop() {
+
+    auto setup_common = [this]<typename MetricDef>() {
+        const auto used_units = [](const size_t current, const size_t max) {
+            return max - current;
```
I think it's useful to return this normalized [0,1], since we don't have a metric for the limit.

IoannisRP replied:
i was thinking of a case where a system is maxed out and then we change the setting.
In the normalized version we would see a dip in utilization, which could either be that requests stop coming or that the setting was changed.
In absolute version we would see it flat-lined at e.g. 500 and then start climbing higher.

BenPope replied:
The question is how would you set an alert? I'm happy to leave it at an absolute value, but also include a metric for the limit.

IoannisRP replied:
> The question is how would you set an alert?

Valid point.

> but also include a metric for the limit.

i think having a normalized value and the limits is the best of both worlds. Easy monitoring on how full your system is and my previous comment goes away.

IoannisRP replied:
i ended up going with just the normalized metrics. It sounds too much to have the maximum values as well, just for an edge case.

BenPope commented on src/v/pandaproxy/server.h:
```diff
@@ -129,13 +132,32 @@ class server {
     bool _has_routes;
     context_t& _ctx;
     json::serialization_format _exceptional_mime_type;
+    probe _probe;
```
`probe` could be forward declared, held by `std::unique_ptr`, started and stopped in `server::start/stop`, and the definition along with the metrics include all moved to the implementation file.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59721-->
<!--2cdf5508-12ea-409c-91a2-88fb2ae79557-->
<!--116c0309-cda9-4809-9583-806befda6198-->
<!--58979cc2-04b4-4252-b720-9df2cd106297-->
<!--title-->
## Retry command for [Build#59721](https://buildkite.com/redpanda/redpanda/builds/59721)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/pandaproxy_test.py::PandaProxyInvalidInputsTest.test_topic_produce_request_too_big
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":false}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--da52b7fa-e8e4-43ff-a0ff-34e6499d281b-->
<!--TEST RESULTS-->
<!--c0342fac-46c4-44ab-9d68-fc9acc410641-->
<!--TEST RESULTS-->
<!--1bea46ff-de9a-43e2-88b6-9e2ce6a2e87f-->
<!--TEST RESULTS-->
<!--701bd620-6419-4611-9bae-5c113662d667-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59721">build#59721</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=20.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59721#0193bffd-6f30-4082-9f96-97ee012061d2 | FAIL | 0/6 |
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59721#0193bffd-6f2e-4cf0-b5d5-732e2009ef7b | FAIL | 0/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59721#0193bffd-6f2f-429e-ba83-7880fd23e05e | FLAKY | 2/6 |
| rptest.tests.pandaproxy_test.PandaProxyInvalidInputsTest.test_topic_produce_request_too_big | ducktape | https://buildkite.com/redpanda/redpanda/builds/59721#0193bffd-6f2d-4362-99bd-847c422c8215 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59721#0193bffd-6f30-4082-9f96-97ee012061d2 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59721#0193bffd-6f30-4082-9f96-97ee012061d2 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59721#0193bffd-6f30-4082-9f96-97ee012061d2 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59721#0193bffd-6f30-4082-9f96-97ee012061d2 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59729">build#59729</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=20.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59729#0193c11a-18ba-4ad7-8a29-7aea1a0d6f63 | FAIL | 0/6 |
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59729#0193c11a-18b8-44a2-8b92-c1ad1f800d14 | FAIL | 0/6 |
| rptest.tests.cloud_storage_scrubber_test.CloudStorageScrubberTest.test_scrubber.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59729#0193c11a-18b7-4d6e-a3f1-0a6aea86eae2 | FLAKY | 5/6 |
| rptest.tests.datalake.compaction_gaps_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59729#0193c11a-18ba-4ad7-8a29-7aea1a0d6f63 | FAIL | 0/1 |
| rptest.tests.maintenance_test.MaintenanceTest.test_maintenance_sticky.use_rpk=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/59729#0193c0fd-d2e3-4a01-8254-bfd0e464fce8 | FAIL | 0/1 |
| rptest.tests.pandaproxy_test.PandaProxyInvalidInputsTest.test_topic_produce_request_too_big | ducktape | https://buildkite.com/redpanda/redpanda/builds/59729#0193c11a-18b7-4d6e-a3f1-0a6aea86eae2 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59729#0193c11a-18ba-4ad7-8a29-7aea1a0d6f63 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=False.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59729#0193c11a-18ba-4ad7-8a29-7aea1a0d6f63 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=False.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59729#0193c11a-18ba-4ad7-8a29-7aea1a0d6f63 | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59729#0193c11a-18ba-4ad7-8a29-7aea1a0d6f63 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59811">build#59811</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59811#0193d07f-3cce-433d-8e7b-22c2ea6e16fa | FAIL | 0/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59848">build#59848</a></summary>


| test_id                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59848#0193d486-cbb6-491c-9fde-409d02d26ddd | FAIL          | 0/6      |
</details>
<!--/body-->

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/a8b2628cec21662cdbdf68faa102e1772a724718..83e1b8dcf337b70f57e69dcae413a4021caf37a1):
- move probe out of server
- change metrics to report normalized values
- removed header in response

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59729-->
<!--7be3ac0e-c715-48bc-b31f-973a7ae8accd-->
<!--b4682f6e-3b47-43cb-91c1-d804f5f8f6c1-->
<!--03712f4e-2e34-4462-8777-3ff4e3b1035b-->
<!--2250b355-79c0-4430-9857-a52c60a5f776-->
<!--title-->
## Retry command for [Build#59729](https://buildkite.com/redpanda/redpanda/builds/59729)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/maintenance_test.py::MaintenanceTest.test_maintenance_sticky@{"use_rpk":false}
tests/rptest/tests/pandaproxy_test.py::PandaProxyInvalidInputsTest.test_topic_produce_request_too_big
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":false,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"enable_failures":true,"mixed_versions":false,"with_iceberg":true,"with_tiered_storage":true}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}
tests/rptest/tests/datalake/compaction_gaps_test.py::CompactionGapsTest.test_translation_no_gaps@{"cloud_storage_type":1}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
```
<!--/body-->

BenPope commented on src/v/pandaproxy/probe.cc:
```diff
@@ -104,4 +104,80 @@ void probe::setup_public_metrics() {
+        defs.reserve(3);
+        defs.emplace_back(
+          sm::make_gauge(
+            "inflight_requests_capacity",
```
Perhaps something like this is clearer:
https://prometheus.io/docs/practices/naming/#base-units
```suggestion
            "inflight_requests_usage_ratio",
```

BenPope commented on src/v/pandaproxy/probe.cc:
```diff
@@ -104,4 +104,80 @@ void probe::setup_public_metrics() {
+                return usage(_ctx.inflight_sem.current(), _ctx.max_inflight);
+            },
+            sm::description(ssx::sformat(
+              "Normalized capacity of requests being executed in parallel for "
```
Maybe something like "Usage ratio of in-flight requests"

BenPope commented on src/v/pandaproxy/probe.cc:
```diff
@@ -104,4 +104,80 @@ void probe::setup_public_metrics() {
+            .aggregate({}));
+        defs.emplace_back(
+          sm::make_gauge(
+            "requests_memory_capacity",
+            [this, usage] {
+                return usage(_ctx.mem_sem.current(), _ctx.max_memory);
+            },
+            sm::description(ssx::sformat(
+              "Normalized memory being used by requests to {}", _group_name)))
```
I think memory usage ratio is a better description

BenPope commented on src/v/pandaproxy/probe.h:
```diff
@@ -12,6 +12,7 @@
 #pragma once
 
 #include "metrics/metrics.h"
+#include "pandaproxy/server.h"
```
I think `server::ctx` can be forward declared.

IoannisRP replied:
`server::context_t` cannot be forward declared because it is an internal struct,.
Given how it is inherited from in the `ctx_server::context_t` and indirectly in the `auth_ctx_server::context_t`,  I don't think it is worth bringing out of the `server` just to break a compilation dependency.

BenPope replied:
Ahh, nothing to do here for this PR, then.

BenPope commented on src/v/pandaproxy/server.cc:
```diff
@@ -240,13 +245,19 @@ ss::future<> server::start(
         co_await _server.listen(addr, cred);
     }
+
+    _probe = std::make_unique<server_probe>(_ctx, _public_metrics_group_name);
```
This should probably be constructed prior to the listeners.

BenPope commented on src/v/pandaproxy/server.cc:
```diff
@@ -240,13 +245,19 @@ ss::future<> server::start(
 }
 
 ss::future<> server::stop() {
+    _probe.reset(nullptr);
```
This should outlive pending requests.

BenPope commented on src/v/pandaproxy/server.h:
```diff
@@ -16,6 +16,7 @@
 #include "hashing/jump_consistent_hash.h"
 #include "hashing/xx.h"
 #include "kafka/client/client.h"
+#include "metrics/metrics.h"
```
I don't think this is required here.

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/83e1b8dcf337b70f57e69dcae413a4021caf37a1..55c390764df86c9bdda2a261967908efd0043a14):
- Probe gets created before listeners.
- Probe gets reset after requests have finished.
- Renamed ratio metrics.
- Fixed dt test to follow updated kafka service memory limits. Also changed test to operate on 256Mb instead of 128Mb to be less restricted.

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/55c390764df86c9bdda2a261967908efd0043a14..1c5c4548ce9d7a6ae65d89c815e0b9f4913d76cf):
- removed a few headers that sneaked in

BenPope commented on src/v/pandaproxy/probe.cc:
```diff
@@ -104,4 +104,78 @@ void probe::setup_public_metrics() {
+            .aggregate({}));
+        defs.emplace_back(
+          sm::make_gauge(
+            "memory_usage_ratio",
```
Is it worth including `inflight_requests` in the name / description?

IoannisRP replied:
Something like
`name : inflight_requests_memory_usage_ratio`
`desc : Usage ratio of memory by inflight requests for {}`
?

BenPope replied:
Added docs as reviewer to check wording.

Deflaimun commented:
Hi guys. I asked other doc writers to take a look. 

While we're here, I see the labels are defined as "shards". How does that work? It's because they have `shard_local_cfg()` attached to them? Or labels are something else entirely? 

Feel free to DM me if you want to talk about it.

BenPope commented:
> While we're here, I see the labels are defined as "shards". How does that work? It's because they have `shard_local_cfg()` attached to them? Or labels are something else entirely?

It's a number representing the logical core that the Redpanda reactor is running on. When Redpanda is given 4 CPU Cores there will be 4 reactors, one on each of the CPU cores, e.g.: `--smp=4` or `-c4`, there will be shard 0 to shard 3.

JakeSCahill commented:
I'll add these metrics to our docs for 24.3 and backport to 24.2: https://redpandadata.atlassian.net/browse/DOC-869

I assume they will go out in the next patch release for each.

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/1c5c4548ce9d7a6ae65d89c815e0b9f4913d76cf..09194538d157167c84aefddafcc2d30db94a0887):
- changed name and description for memory metric

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59811-->
<!--6c69a9b5-e0cb-486c-9571-074c64952637-->
<!--title-->
## Retry command for [Build#59811](https://buildkite.com/redpanda/redpanda/builds/59811)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy@{"cloud_storage_type":2,"short_retention":true}
```
<!--/body-->

BenPope commented on src/v/pandaproxy/probe.cc:
```diff
@@ -104,4 +104,79 @@ void probe::setup_public_metrics() {
+                return usage(_ctx.mem_sem.current(), _ctx.max_memory);
+            },
+            sm::description(ssx::sformat(
+              "Usage ratio of memory by inflight requests for {}",
```
This sounds like it's the proportion of total memory used by inflight requests. Perhaps
```suggestion
              "Usage ratio of memory for inflight requests for {}",

BenPope replied:
Or maybe:
```suggestion
              "Memory usage ratio of inflight requests in the {}",

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/09194538d157167c84aefddafcc2d30db94a0887..067fc7b90da47a2c24d6eb96abc1ac0925610cf0):
- reworded semaphore metric descriptions

BenPope reviewed (approved):
Please fix the cover letter to reflect the latest changes.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59848-->
<!--a49a924a-d253-46a7-8729-daeaa228455b-->
<!--title-->
## Retry command for [Build#59848](https://buildkite.com/redpanda/redpanda/builds/59848)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
```
<!--/body-->

IoannisRP commented:
CI Failures:
- [CORE-8573](https://redpandadata.atlassian.net/browse/CORE-8573)

[CORE-8573]: https://redpandadata.atlassian.net/browse/CORE-8573?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

michael-redpanda commented on src/v/pandaproxy/probe.cc:
```diff
@@ -104,4 +104,79 @@ void probe::setup_public_metrics() {
+            },
+            sm::description(ssx::sformat(
+              "Usage ratio of in-flight requests in the {}", _group_name)))
+            .aggregate({}));
```
This means we won't ever aggregate per shard right?  @BenPope is this acceptable?

BenPope replied:
> This means we won't ever aggregate per shard right?  @BenPope is this acceptable?

Yeah, you want to be able to see whether it's balanced, and ratios don't aggregate well.

mike-boquard replied:
Ok fair enough

michael-redpanda reviewed (commented):
One question about aggregation of metrics

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
/backport v24.1.x

dotnwat reviewed (commented):
👍
